### ISSUE TITLE
Iteration # 1-offset in DDP logging

### ISSUE THREAD

**Issue by QihanW** [pytorch/pytorch#79464]
### 🐛 Describe the bug

When using the _get_ddp_logging_data() to collect logging data, the iteration# has 1-offset. In the first iteration, ddp logging cannot collect any data. In the next iterations, the ddl logging data are always 1 iteration offset. 

Here are the results to compare the ddl logging data with the profiler traces:

Iteration \ Time (ms) | Forward_compute_time (DDP) | Forward time in profiling
-- | -- | --
1 | 1953.336832 | 20.661
2 | 20.523072 | 22.753
3 | 23.968768 | 28.798
4 | 29.75024 | 18.042
5 | 18.861792 | 40.775
6 | 41.697664 |  

</div></b>

Here is the sample code:

      import os
      import socket
      import sys
      import logging
      import time
      from collections import OrderedDict
      import torch
      import torch.distributed as dist
      import torch.nn as nn
      from torch.autograd.profiler import record_function
      from torch.distributed.optim import PostLocalSGDOptimizer
      import torchvision
      from torchvision import models, transforms
  
      def run_worker(rank, world_size):

        torch.cuda.set_device(rank)
        dist.init_process_group("nccl", rank=rank, world_size=world_size)
  
        model = models.resnet18(pretrained=False).cuda()
        ddp_model = nn.parallel.DistributedDataParallel(
            module=model,
            device_ids=[rank],
        )
        # Define loss function, input, labels, and optimizer.
        loss_fn = nn.MSELoss()
        input = torch.randn(10, 3, 224, 224).cuda()
        labels = torch.randn(10, 1000).cuda().to(rank)
        optimizer = torch.optim.SGD(ddp_model.parameters(), lr=0.001)
        
        with torch.profiler.profile(
            activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
            on_trace_ready=torch.profiler.tensorboard_trace_handler("/tmp/traces"),
            record_shapes=True,
        ) as prof:
            for i in range(0, 10):
                with record_function(f"#Step {i}"):
                    optimizer.zero_grad()
                    with record_function("#forward"):
                        output = ddp_model(input)
                    loss = loss_fn(output, labels).half().cuda()
                    with record_function("#backward"):
                        loss.backward()
                    optimizer.step()
                    ddp_logging = ddp_model._get_ddp_logging_data()
                    print(ddp_logging)
        # Cleanup.
        dist.destroy_process_group()
      
      if __name__ == "__main__":
        os.environ["MASTER_ADDR"] = "localhost"
        os.environ["MASTER_PORT"] = "12345"
        world_size = torch.cuda.device_count()
        mp.spawn(run_worker, args=(world_size,), nprocs=world_size, join=True)`

### Versions

since pytorch 1.9

**Comment by zhaojuanmao** [comment]
thanks for reporting, will take a look at it
